smartcontractkit/external-adapters-js
# Reduce GHA workflow check workload, optimize dependency caching

**Currently not ready for deployment until there's a mechanism to build composite adapters when a source adapter it depends on changes.**

This is a tweak to CI that cuts build times by about 3/4~4/5 by having certain workflows run against changed adapters instead of all adapters.

Affected workflows/steps:
* main.yml/run-documentation-check (yarn generate:readme)
* main.yml/build-adapters (build a limited set)
* generate-readmes.yml
* release.yml/publish-artifacts

Note, the main purpose of this PR is to save time on checks against develop, but it should also be safe to push to the production workflows, generate-readmes.yml and release.yml. See [this comment](https://github.com/smartcontractkit/external-adapters-js/pull/2407#issuecomment-1347422707) for details.

Here's what this PR does
* By default, the build matrix JS script (scripts/gha) will only present adapters that have changed in the source branch, based on filepaths from git diff. This affects all workflows using build matrixes (main and release).
* By default, changes generate:readme to only generate readmes for changed adapters, which is done through a bash script. This affects main and generate-readmes.
* If either of the above need to be bypassed, manual workflow runs have a BUILD_ALL parameter to bypass the filtering behavior. All of the affected workflows can now be run manually if they couldn't before.
* Regardless of whether the BUILD_ALL flag is passed or not, all workflows will run against all adapters if any changes to core/ or scripts/ appears in the diff.

How to test these changes:
1. Check out a new branch based on **this branch (feature/PDI-188)**  (calling the newly created branch `example` below)
2. In the example branch
        1. Make a change to any source, composite, non-deployable, example, or observable adapter
        2. Push that change
        3. Create a PR against this branch (feature/PDI-188)
        4.  Observe that Documentation Test and Test and Build steps only operate on your changed adapter
        5. Make any change to core or scripts
        6. Push that change
        7. Observe that Documentation Test and Test and Build run against all adapters

You can also test locally by creating a new branch based of this branch, changing files, and then running one or both of the following commands (alter UPSTREAM_BRANCH and BUILD_ALL to your liking):
* `source .github/get-changed-adapters-array.sh && ADAPTERS=$(getChangedAdapterArray) && UPSTREAM_BRANCH=develop getChangedAdapterArray`
* `BUILD_ALL=false yarn generate:gha:matrix`

See this draft commit to see this work in action (you can't see it in this branch because a change to packages/scripts will always build all): https://github.com/smartcontractkit/external-adapters-js/pull/2394
